source/common/json/json_internal.cc

```diff
@@ -785,9 +785,13 @@ ObjectSharedPtr Factory::loadFromProtobufStruct(const ProtobufWkt::Struct& proto
   return loadFromProtobufStructInternal(protobuf_struct);
 }
 
-std::string Factory::serialize(absl::string_view str) {
+std::string Factory::serialize(absl::string_view str, bool replace_valid_utf8) {
   nlohmann::json j(str);
-  return j.dump();
+  // Call with defaults except in the case of UTF-8 errors which we may replace
+  // invalid UTF-8 characters instead of throwing an exception.
+  return j.dump(-1, ' ', false,
```

Comment:
actually it doesn't look too hard to just insert the unicode escapes here here. Should we do that?

Reply:
I am inclined don't to introduce additional complexity. nlohmann/json is widely used (may be more widely be used than the envoy, haha).
So, if this behavior is OK to the users of nlohmann/json, I think it's also OK to envoy users.

Reply:
I don't buy this logic at all per chat.

Is it a problem to just keep the escaping logic in the handler you wrote? Then we actually can drop the exception-generating version completely, simplifying the code and tests without changing the contract.

Reply:
We cannot get agreement on every point because we are to independent individuals, but I think it doesn't matter, right?

> Is it a problem to just keep the escaping logic in the handler you wrote? Then we actually can drop the exception-generating version completely, simplifying the code and tests without changing the contract.

I actually hope I can define a custom error handler, then I can copy your escape logic directly. But now, the error handler not an extendable handler, by is an `enum` that provided by the nlohmann/json. And only `strict`, `ignore`, and `replace` could be used.

(If we still use rapidjson. Thing wouldn't be so complex. Although it's too late to talking about that. orz)